### DISCUSSION TITLE
How to automatically generate allow-lists from a JS/TS codebase?

### DISCUSSION THREAD

**Discussion by dminkovsky** [hasura/graphql-engine#6078] [General] [answered]
I would like to use allow-lists, but entering them to Hasura by hand does not seem feasible. Has anyone explored any automated methods? I am thinking something that scans my code for `gql` tags (I am using Apollo Client) and produces an allow list.

**Comment by dminkovsky** [comment]
Looks like apollo-tool's [client:extract](https://github.com/apollographql/apollo-tooling#apollo-clientextract-output) will do the job. Can be paired with a script that pushes to the metadata API to fully automate, or can be uploaded to Hasura by hand.

**Reply by msimon** [reply]
@dminkovsky I just apply some fixes to this lib that does that for you entirely: https://github.com/TallerWebSolutions/hasura-allowed-queries.

The fixes are in PR at the moment, but you can run the dev version from my fork: https://github.com/msimon/hasura-allowed-queries

**Reply by dminkovsky** [reply]
@msimon **Thank you!** I will check these out.

**Reply by dminkovsky** [reply]
@msimon by the way, have you had any experience with versioning? Let's say I make a change to a query, but I have clients in the wild using old versions of that query—any thoughts on how those clients old could keep working if I update the allow-list?

**Reply by msimon** [reply]
Sadly the allow lists are all or nothing. So you will have to at least know the past query.

If you have a client in the wild that uses old queries and can know which query they are using, you can have an add-only approach to it.

For example, I currently uses:
`yarn dev -h http://localhost:8080 -s local_admin_secret -p '../../frontend/**/*.graphql' -r`

The `-r` option will erase all the allow lists before inserting the ones that are in the frontend folder.  But if you remove the `-r`, it will only add or update queries. As long as you're creating queries with a new name, you will be fine.

You can then manually delete the ones that you are not interested in anymore directly in Hasura.

**Reply by msimon** [reply]
I think it could be even interesting to add the feature to only add and generate new allow list query names in case of conflict. That would especially be useful for mobile app where several clients are always in the wild.

**Reply by msimon** [reply]
Actually I realise that I will need something like that in some cases. I'm working on a solution.

**Reply by dminkovsky** [reply]
@msimon cool, thanks again for posting and for the update. Would love to see what you come up with. My bandwidth is totally saturated right now, but this will be important to me soon.

**Reply by msimon** [reply]
@dminkovsky push to my repo.

I added the option `-v, --version <version>` that allows several versions. Once it finds an updated query, it will create a new version of it.  At the moment, it never deletes anything.
So some kind of manual cleanup is necessary. I may add automatic cleanup as an option later if needed. Maybe based on the number of days or list of supported versions.

I personally using it this way: 
```
yarn dev -h http://localhost:8080 -s local_admin_secret -p '../../frontend/**/*.graphql' -v ${GIT_VERSION}
```

This will create allow list that respect this format: `($NAME__($TIMESTAMP_$VERSION)`.  `loggedUser___(1624536406478-0dfe607)`

**Reply by dminkovsky** [reply]
Thank you so much for sharing, Marc. I will definitely have a look when I get to this.

**Reply by dminkovsky** [reply]
I wonder what the performance implications of the allow-list are, especially of there are many queries. I really wish Hasura had hashed/persisted queries.

**Reply by msimon** [reply]
I haven't checked their code, but I would not be surprised it's a hash.
The allow-list queries are stored in memory into their metadata, and they have to match exactly. Even parameters order is required.
So it's really just a plain string comparison, which is certainly very fast.

**Comment by msimon** [comment, marked as answer]
@dminkovsky I ended up packaging and releasing it on my own as it does not seems like TallerWebSolutions is active anymore.
You can find the updated doc here: https://github.com/HomeskilletHealthInc/hasura-allow-list-manager
and package here: https://www.npmjs.com/package/@homeskillet/hasura-allow-list-manager

I have been using for several deployment so far, and it works very well :)

**Reply by dminkovsky** [reply]
@msimon thanks so much for this package and the README. I look forward to trying it, which will hopefully mean that my app has become successful enough to attract negative attention 😜. Soooon 🤞.

**Reply by dminkovsky** [reply]
Hey Marc looks like this also exists: https://www.graphql-code-generator.com/plugins/other/hasura-allow-list

**Reply by msimon** [reply]
Nice!

Have you tried it yet? Does it support the versioning by any chance?

**Reply by dminkovsky** [reply]
Haven't tried it yet. What do you mean by versioning?

**Reply by msimon** [reply]
So one of the issues that I had to address is that I apply the allow-list and the rest of the hasura configuration during deployment.
To be able to support backward compatibility, you must be able to allow both the past version of the allow-list's query and the new one.

If you are only having a website, that can be fine as the deployment can be pretty quick. Still not optimal though, as you still have delay where request will fail and will need to force the clients to reload themselves before making the query.
But if you have a client app, it can takes several days for people to update. So you definitely need several versions of allow-list query, or the client app will simply not be able to make some requests before the app get updated.

**Reply by msimon** [reply]
The current support in my repo isn't optimal though. When using versioning, you can only add to the allow-list.
I haven't yet work on removing past version when needed. So if you had a vulnerability at some point, you'll have to edit the file manually to remove it.

**Reply by msimon** [reply]
I noticed I actually explained how it works here haha:
https://github.com/HomeskilletHealthInc/hasura-allow-list-manager#versioning-behavior

**Reply by dminkovsky** [reply]
Oh well yeah I would think you could do versioning by checking in the generated allow-lists that are generated by this tool, or something like that?

**Reply by dminkovsky** [reply]
Of course not breaking old clients is a huge consideration...

**Reply by msimon** [reply]
> Oh well yeah I would think you could do versioning by checking in the generated allow-lists that are generated by this tool, or something like that?

Mmm I don't know how that would work. Last time I check, you could only set the allow-list file at once.
So if you have versioning, i think you'll have to add it to this file your uploading. Maybe Hasura's api change since then though

**Reply by dminkovsky** [reply]
My bad I haven't still actually used allowed-lists so I was a bit confused about them. I just looked at the docs and I see what you mean. I'll try this plugin out because I'm using this codegen util already (it's really nice, generates `TypePolicies` for example!), and I'll update here about how versioning works (or does not work :))

**Reply by dminkovsky** [reply]
Oh and by the way, looks like [the other answer](https://github.com/hasura/graphql-engine/discussions/6078#discussioncomment-2112947) is this plugin.

**Comment by BenoitRanque** [comment]
We have released a plugin for graphql codegen that allows you to generate the metadata file you need for this.
You can find the pull request at https://github.com/dotansimha/graphql-code-generator/issues/7476
If this is a feature you'd like, please show support on the issue so it gets accepted! Also interested to hear any feedback.

**Reply by dminkovsky** [reply]
I came here to post this :) https://www.graphql-code-generator.com/plugins/other/hasura-allow-list. Awesome!

**Reply by dminkovsky** [reply]
@praveenweb this should be the accepted answer?
